### PR TITLE
feat(ui-shell): add experimental platform header

### PR DESCRIPTION
**Pull request by joshblack**
Introduces an experimental platform header. This is meant as an initial PR so that the diff doesn't get overwhelming. Specifically, this PR is looking to add in the basic UI Shell header with some token support and inlined experimental styles for things like type while we get that sorted out.

This is also a first chance for design to weigh in on implementation and offer feedback for things we need to fix with this first pass!

#### Changelog

**New**

- `components/ui-shell` folder where we will keep the UI Shell-specific experimental work for now
  - `platform-header` initially introduced, though it is not complete by any means

**Changed**

**Removed**

#### Testing / Reviewing

* Make sure to switch the `ui-shell` flag to `true` when reviewing!

[URL Preview](https://carbon-dev-environment-nonlacteous-lich.mybluemix.net/component/ui-shell--header)

### CODE REVIEW

File: src/components/ui-shell/_variables.scss [hunk 1]
```diff
@@ -0,0 +1 @@
+$unit: 8px;
```
Comment (asudoh) [inline src/components/ui-shell/_variables.scss]: Should we use `rem` to accommodate zooming?

File: src/components/ui-shell/ui-shell--header.hbs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    <svg width="20" height="20" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 32 32" aria-hidden="true">
+      <path d="M8.24 25.14L7 26.67a14 14 0 0 0 4.18 2.44l.68-1.88a12 12 0 0 1-3.62-2.09zm-4.05-7.07l-2 .35A13.89 13.89 0 0 0 3.86 23l1.73-1a11.9 11.9 0 0 1-1.4-3.93zm7.63-13.31l-.68-1.88A14 14 0 0 0 7 5.33l1.24 1.53a12 12 0 0 1 3.58-2.1zM5.59 10L3.86 9a13.89 13.89 0 0 0-1.64 4.54l2 .35A11.9 11.9 0 0 1 5.59 10zM16 2v2a12 12 0 0 1 0 24v2a14 14 0 0 0 0-28z" />
+    </svg>
+  </button>
+  <a class="bx--platform-header__name" href="#" title="Platform home">
+    IBM [Platform]
+  </a>
+  <nav class="bx--platform-header__nav">
+    <ul class="bx--platform-header__links">
+      <li class="bx--platform-header__link-item">
```
Comment (asudoh) [inline src/components/ui-shell/ui-shell--header.hbs]: Great to see Handlebars `{{#each}}` directive used in this file - Ref: https://github.com/IBM/carbon-components/pull/997

**Review by asudoh — COMMENTED** [review]
LGTM except for a couple - Thanks @joshblack!

File: src/components/ui-shell/_variables.scss (hunk 1 above)
Comment (joshblack) [reply src/components/ui-shell/_variables.scss]: Up for anything! Whatever works best 😄

File: src/components/ui-shell/ui-shell--header.hbs (hunk 2 above)
Comment (joshblack) [reply src/components/ui-shell/ui-shell--header.hbs]: Can definitely dive into the configuration, maybe I'll add that in another PR when we have sub-menus for the links? That feels complex enough to warrant starting to add it in.

Comment (asudoh) [reply src/components/ui-shell/ui-shell--header.hbs]: Yeah doing that in subsequent change will be good - Thanks @joshblack!

File: src/components/ui-shell/_variables.scss [hunk 3]
```diff
@@ -0,0 +1,4 @@
+// Use rem here to accomodate zoom behavior
+// Reference:
+// https://github.com/IBM/carbon-components/pull/1032#discussion_r210418579
+$unit: 0.5rem;
```
Comment (asudoh) [inline src/components/ui-shell/_variables.scss:4]: Thank you for the change here! Do you want to use `rem()` mix-in in `_typography.scss`?

Comment (joshblack) [reply src/components/ui-shell/_variables.scss:4]: Do you think it's necessary? I typically have only used that for sizes that I'd need a calculator to figure out the px value 😂 

Or is this more-so about the `units` function?

Comment (asudoh) [reply src/components/ui-shell/_variables.scss:4]: This is more for coding convention - I tend to look at how the codebase is written when I jump into project and replicate the coding convention to the code I write. Just wondered if you want to do the same. If you are not feeling like making the change it’ll be no problem!

File: src/components/ui-shell/_platform-header.scss [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+      border: units(0.5) solid transparent;
+    }
+
+    a.#{$prefix}--platform-header__link,
+    a.#{$prefix}--platform-header__link:hover {
+      color: $shell-header-text-01;
+    }
+
+    a.#{$prefix}--platform-header__link:active {
+      background-color: #3d3d3d;
```
Comment (alisonjoseph) [inline src/components/ui-shell/_platform-header.scss]: Is there are reason only some of the colors are in the `theme.scss` file?

**Review by alisonjoseph — COMMENTED** [review]
Overall looks good! Just a couple comments/questions from me. 

I noticed in Safari the focus state doesn't show up on the links. Seems fine in all other browsers. 

Do we have a responsive/mobile/small screen design for this?

 Is there no "click/down" state color change on the links on purpose? I noticed that the icon background turn blue on click. Maybe more of a design question?

File: src/components/ui-shell/_platform-header.scss (hunk 4 above)
Comment (joshblack) [reply src/components/ui-shell/_platform-header.scss]: Still waiting to hear back from Anna on if there are hover/active tokens, or not, for these!

**Comment by joshblack** [comment]
@alisonjoseph added in support for the focus state on links in the header, no responsive designs at the moment.

**Comment by aagonzales** [comment]
Looking good! Just two things:

- [x]  The links should be fixed to left, at 272px from the left edge.
- [ ] ~Once the links meet the icon headers the links should drop into a hamburger menu to the on the left side of the header.~
image

**Comment by joshblack** [comment]
Thanks @aagonzales! Will make the update for the margin, the responsive chunk might be better in a future PR 👍

**Comment by joshblack** [comment]
cc @alisonjoseph @asudoh do you have a chance to re-review?

**Review by alisonjoseph — APPROVED** [review]
This looks good to me, my only questions were around  whether or not there should be color tokens for hover/active colors.

**Comment by joshblack** [comment]
@alisonjoseph I'll add them in real quick 👍

File: src/components/ui-shell/ui-shell-preview.hbs [hunk 5]
```diff
@@ -0,0 +1,25 @@
+<!DOCTYPE html>
```
Comment (asudoh) [inline src/components/ui-shell/ui-shell-preview.hbs:1]: Do you want to move this to `demo/views/layouts` so other components can use this preview potentially…?

File: src/components/ui-shell/_theme.scss [hunk 6]
```diff
@@ -0,0 +1,33 @@
+// Header bar background
+$shell-header-bg-01: #252525;
```
Comment (asudoh) [inline src/components/ui-shell/_theme.scss:2]: @IBM/carbon-designers I believe we’ll create theme tokens that applies to system-wide and go away from feature-specific color tokens in future - Is it correct?

File: src/components/ui-shell/ui-shell.hbs [hunk 7]
```diff
@@ -0,0 +1 @@
+{{> platform-header }}
```
Comment (asudoh) [inline src/components/ui-shell/ui-shell.hbs:1]: Nitpick - Do you want to remove the whitespace right before `}`?

**Review by asudoh — COMMENTED** [review]
Great to see more progress @joshblack!

File: src/components/ui-shell/_theme.scss (hunk 6 above)
Comment (joshblack) [reply src/components/ui-shell/_theme.scss:2]: Ideally we'd use the tokens that we have already for colors, but for the shell work we have this set of tokens for now (can check in our squad channel to see Anna's post with this set).

File: src/components/ui-shell/ui-shell.hbs (hunk 7 above)
Comment (joshblack) [reply src/components/ui-shell/ui-shell.hbs:1]: Just was following the convention here: https://handlebarsjs.com/partials.html

```hbs
{{> myPartial }}
```

File: src/components/ui-shell/ui-shell-preview.hbs (hunk 5 above)
Comment (joshblack) [reply src/components/ui-shell/ui-shell-preview.hbs:1]: Maybe. Would rather wait to move it when the need is there than move it prematurely.

**Review by asudoh — APPROVED** [review]
Signing off putting 
https://github.com/IBM/carbon-components/pull/1032/files#r210429800 aside given it’s minor - Also great to see the thing about layout template location address in near future, esp. given no layout template has been put under `src` in this codebase. Thanks @joshblack!

**Comment by alisonjoseph** [comment]
This seems good to merge as long as design signs off 😸

**Pull request merged**